### PR TITLE
(maint) Make dockerfile configurable, and refactor PuppetDockerTools::Run to be a class

### PR DESCRIPTION
**Pull request by underscorgan**

### CODE REVIEW

**Comment by underscorgan** [comment]
Added tests for building with a different dockerfile, and updated the image build to actually use the custom dockerfile if passed.

File: bin/puppet-docker [hunk 1]
```diff
@@ -56,27 +58,33 @@ if options['IMAGE'].nil?
...
+      command_runner.rev_labels
+    elsif options['spec']
+      command_runner.spec
+    elsif options['test']
+      command_runner.lint
+      command_runner.spec
+    elsif options['version']
+      command_runner.version
+    elsif options['help']
+      puts doc
```
Comment (mwaggett) [inline bin/puppet-docker]: I think this is redundant with the first if?

Comment (underscorgan) [reply bin/puppet-docker]: updated

File: lib/puppet_docker_tools/runner.rb [hunk 2]
```diff
@@ -128,57 +92,38 @@ def push(directory, repository: , namespace: )
...
-    # @param directory The directory containing the Dockerfile to update
-    # @param namespace The namespace for the labels
-    def rev_labels(directory, namespace: )
-      dockerfile = File.join(directory, PuppetDockerTools::DOCKERFILE)
-
-      unless File.exist? dockerfile
-        fail "File #{dockerfile} doesn't exist."
-      end
+    def rev_labels
+      file = File.join(directory, dockerfile)
```
Comment (mwaggett) [inline lib/puppet_docker_tools/runner.rb]: Do we still want to check that this file exists?

File: spec/lib/puppet_docker_tools/runner_spec.rb [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+  describe '#spec' do
+    it "runs tests under the 'spec' directory" do
+      tests=["/tmp/test-image/spec/test1_spec.rb", "/tmp/test-dir/spec/test2_spec.rb"]
+      expect(Dir).to receive(:glob).with("/tmp/test-image/spec/*_spec.rb").and_return(tests)
+      expect(RSpec::Core::Runner).to receive(:run).with(tests, $stderr, $stdout).and_return(nil)
+      runner.spec
+    end
+  end
+end
```
Comment (mwaggett) [inline spec/lib/puppet_docker_tools/runner_spec.rb]: I assume this whole file was just a copy paste from the old spec file (plus changing class names where appropriate)?

File: lib/puppet_docker_tools/runner.rb [hunk 4]
```diff
@@ -6,25 +6,31 @@
...
-    #        the image from
-    # @param repository The repository this image will be pushed to
-    # @param namespace The namespace for the version label in the dockerfile
     # @param no_cache Whether or not to use existing layer caches when building
     #        this image. Defaults to using the cache (no_cache = false).
-    def build(directory, repository: , namespace: ,no_cache: false)
+    def build(no_cache: false)
       image_name = File.basename(directory)
-      version = PuppetDockerTools::Utilities.get_value_from_env('version', namespace: namespace, directory: directory)
+      version = PuppetDockerTools::Utilities.get_value_from_env('version', namespace: namespace, directory: directory, dockerfile: dockerfile)
```
[the remark this answers is not in the document]
Comment (underscorgan) [reply lib/puppet_docker_tools/runner.rb]: nope, there are attr_accessors for those

File: lib/puppet_docker_tools/runner.rb (hunk 2 above)
Comment (underscorgan) [reply lib/puppet_docker_tools/runner.rb]: that was moved to the instance instantiation.

File: spec/lib/puppet_docker_tools/runner_spec.rb (hunk 3 above)
Comment (underscorgan) [reply spec/lib/puppet_docker_tools/runner_spec.rb]: Some of it was a copy/paste, some was additions. Definitely recommend reviewing each commit separately.

File: lib/puppet_docker_tools/runner.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (mwaggett) [reply lib/puppet_docker_tools/runner.rb]: mmkay. yeah, I was trying to google to see what exactly the attr_accessors give you, but I was having trouble finding examples where the instance variables were used inside of class methods *without* `@`. Most of the stuff I found were of the `Foo.name` variety. if that makes sense.

File: lib/puppet_docker_tools/runner.rb (hunk 2 above)
Comment (mwaggett) [reply lib/puppet_docker_tools/runner.rb]: ahhhh I see 👍

File: spec/lib/puppet_docker_tools/runner_spec.rb [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      allow(container).to receive(:tap).and_return(container)
+      allow(container).to receive(:attach)
+      allow(container).to receive(:wait)
+      allow(container).to receive(:logs).and_return('container logs')
+    end
+
+    it "should lint the container" do
+      allow(container).to receive(:json).and_return(passing_exit)
+      runner.lint
+    end
```
Comment (mwaggett) [inline spec/lib/puppet_docker_tools/runner_spec.rb]: What exactly is being tested here? I don't see any `expect`s.

Comment (underscorgan) [reply spec/lib/puppet_docker_tools/runner_spec.rb]: it's testing that it doesn't raise an exception

Comment (mwaggett) [reply spec/lib/puppet_docker_tools/runner_spec.rb]: ooooh gotcha

**Pull request merged**